### PR TITLE
nixos/patroni: init

### PR DESCRIPTION
**Pull request by phfroidmont**
###### Description of changes
Add a new module for the Patroni service. 
Uses a settings style configuration option.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/cluster/patroni/default.nix
Comment (pennae) [inline nixos/modules/services/cluster/patroni/default.nix:37]:
```suggestion
      description = mdDoc ''
        List of PostgreSQL plugins. PostgreSQL version for each plugin should
        match version for `services.postgresql.package` value.
      '';
```
module docs should preferentially be written in markdown syntax, and ideally all docs should be marked as markdown even if they don't contain docbook tags

File: nixos/modules/services/cluster/patroni/default.nix
Comment (pennae) [inline nixos/modules/services/cluster/patroni/default.nix:102]:
```suggestion
      type = types.listOf types.string;
```

File: nixos/modules/services/cluster/patroni/default.nix
Comment (pennae) [inline nixos/modules/services/cluster/patroni/default.nix:54]:
```suggestion
      type = types.port;
```
same for other port options.

File: nixos/modules/services/cluster/patroni/default.nix
Comment (pennae) [inline nixos/modules/services/cluster/patroni/default.nix:195]: this ought to be configurable, not hardcoded. reusing the postgres uid also clashes with running a non-HA database on the same host for any reason. `isSystemUser` should be set for service users, shells and homedirs should not be set unless absolutely necessary

File: nixos/modules/services/cluster/patroni/default.nix
Comment (pennae) [inline nixos/modules/services/cluster/patroni/default.nix:233]: this looks extremely sketchy. what if a hardware watchdog exists? what if other services also want to use it?

File: nixos/modules/services/cluster/patroni/default.nix
Comment (pennae) [inline nixos/modules/services/cluster/patroni/default.nix:241]: what is this needed for?

Comment (phfroidmont) [reply nixos/modules/services/cluster/patroni/default.nix:241]: It provides `syncobj_admin` which is useful to manage RAFT, although it makes no sense to include it if RAFT is not enabled, I'll change that.

File: nixos/modules/services/cluster/patroni/default.nix
Comment (phfroidmont) [reply nixos/modules/services/cluster/patroni/default.nix:233]: I wanted to provide this for convenience as I suppose this is the most common setup.
I can either remove the `watchdog` setting or rename it to `softwareWatchdog` and be explicit in the doc about what it does.

File: nixos/modules/services/cluster/patroni/default.nix
Comment (phfroidmont) [reply nixos/modules/services/cluster/patroni/default.nix:195]: I basically copied this from the PostgreSQL module.
It would be nice for this to be configurable but at the same time I like the ability to do `sudo -u postgres psql`  (maybe this is why it is done this way in the PostgreSQL module).
Either way I don't mind changing this but I'm not sure what to do exactly. Maybe creating `patroni` as a service user and let people create `postgres` user on their own if they want to?

File: nixos/modules/services/cluster/patroni/default.nix
Comment (phfroidmont) [reply nixos/modules/services/cluster/patroni/default.nix:54]: Will do.

File: nixos/modules/services/cluster/patroni/default.nix
Comment (phfroidmont) [reply nixos/modules/services/cluster/patroni/default.nix:37]: Will do.

File: nixos/modules/services/cluster/patroni/default.nix
Comment (pennae) [reply nixos/modules/services/cluster/patroni/default.nix:195]: one common solution is to have default settings for a number of service accounts (eg `patroni`, `patroni-postgres`, etc) that are created unless the defaults are changed, and used in the configuration where a user/group is required. for this you probably don't need the fixed uid assignment, so merely adding a few user options to the config and creating them unless changed from the default should be quite good.

File: nixos/modules/services/cluster/patroni/default.nix
Comment (pennae) [reply nixos/modules/services/cluster/patroni/default.nix:233]: not sure what the best course of action would be here. maybe split the option in two, one for *using* the watchdog and for setting up the software watchdog device. a note on how to configure a pre-existing watchdog for use with this service could also be helpful.

Comment (phfroidmont) [reply nixos/modules/services/cluster/patroni/default.nix:233]: I dont know how you would configure a pre-existing watchdog, Patroni [documentation](https://patroni.readthedocs.io/en/latest/watchdog.html#setting-up-software-watchdog-on-linux) doesn't explain that.
The documentation also states that *"watchdogs are only supported using Linux watchdog device interface"* so a single option to either use it or not seems fine.
It is also the first time I'm exposed to the concept of a watchdog so I'm not really qualified to give advice here.

Comment (pennae) [reply nixos/modules/services/cluster/patroni/default.nix]: what's in the PR now will work fine. totally missed that `settings` itself references the watchdog, so using a hardware watchdog could just add those settings and leave `softwareWatchdog` unset.

**Comment by phfroidmont** [comment]
I added myself as maintainer if that's alright.
Since I'm going to use this in production I have to do the work anyway.

**Review by pennae — CHANGES_REQUESTED** [review]
more module maintainers is very appreciated :D

looks like the patroni test is failing with odd python errors:
```
node2 # [   62.310696] patroni-start[1060]: Traceback (most recent call last):
node2 # [   62.312397] patroni-start[1060]:   File "/nix/store/09x7b9dnxix6254ry2wlmf9dsiqq9za9-patroni-2.1.3/bin/.patroni-wrapped", line 9, in <module>
node2 # [   62.317008] patroni-start[1060]:     sys.exit(main())
node2 # [   62.319381] patroni-start[1060]:   File "/nix/store/09x7b9dnxix6254ry2wlmf9dsiqq9za9-patroni-2.1.3/lib/python3.9/site-packages/patroni/__main__.py", line 143, in main
node2 # [   62.324128] patroni-start[1060]:     return patroni_main()
node2 # [   62.326276] patroni-start[1060]:   File "/nix/store/09x7b9dnxix6254ry2wlmf9dsiqq9za9-patroni-2.1.3/lib/python3.9/site-packages/patroni/__main__.py", line 135, in patroni_main
node2 # [   62.330734] patroni-start[1060]:     abstract_main(Patroni, schema)
node2 # [   62.333194] patroni-start[1060]:   File "/nix/store/09x7b9dnxix6254ry2wlmf9dsiqq9za9-patroni-2.1.3/lib/python3.9/site-packages/patroni/daemon.py", line 98, in abstract_main
node2 # [   62.337358] patroni-start[1060]:     controller = cls(config)
node2 # [   62.339383] patroni-start[1060]:   File "/nix/store/09x7b9dnxix6254ry2wlmf9dsiqq9za9-patroni-2.1.3/lib/python3.9/site-packages/patroni/__main__.py", line 25, in __init__
node2 # [   62.343567] patroni-start[1060]:     self.dcs = get_dcs(self.config)
node2 # [   62.349066] patroni-start[1060]:   File "/nix/store/09x7b9dnxix6254ry2wlmf9dsiqq9za9-patroni-2.1.3/lib/python3.9/site-packages/patroni/dcs/__init__.py", line 97, in get_dcs
node2 # [   62.355152] patroni-start[1060]:     return item(config[name])
node2 # [   62.357122] patroni-start[1060]:   File "/nix/store/09x7b9dnxix6254ry2wlmf9dsiqq9za9-patroni-2.1.3/lib/python3.9/site-packages/patroni/dcs/raft.py", line 269, in __init__
node2 # [   62.364210] patroni-start[1060]:     self._sync_obj = KVStoreTTL(ready_event.set, self._on_set, self._on_delete, commandsWaitLeader=False, **config)
node3 # [   62.350843] patroni-start[798]: 2022-08-07 14:02:13,549 INFO: no action. I am (node3), the leader with the lock
node2 # [   62.371232] patroni-start[1060]:   File "/nix/store/09x7b9dnxix6254ry2wlmf9dsiqq9za9-patroni-2.1.3/lib/python3.9/site-packages/patroni/dcs/raft.py", line 124, in __init__
node2 # [   62.375213] patroni-start[1060]:     super(KVStoreTTL, self).__init__(self_addr, partner_addrs, conf)
node2 # [   62.379162] patroni-start[1060]:   File "/nix/store/09x7b9dnxix6254ry2wlmf9dsiqq9za9-patroni-2.1.3/lib/python3.9/site-packages/patroni/dcs/raft.py", line 71, in __init__
node2 # [   62.386234] patroni-start[1060]:     super(DynMemberSyncObj, self).__init__(selfAddress, partnerAddrs, conf, transportClass=_TCPTransport)
node2 # [   62.393179] patroni-start[1060]:   File "/nix/store/pcpg66fjsn7vah409hynkhirsjphxdya-python3.9-pysyncobj-0.3.11/lib/python3.9/site-packages/pysyncobj/syncobj.py", line 161, in __init__
node2 # [   62.396355] patroni-start[1060]:     self.__raftLog = createJournal(self.__conf.journalFile)
node2 # [   62.402199] patroni-start[1060]:   File "/nix/store/pcpg66fjsn7vah409hynkhirsjphxdya-python3.9-pysyncobj-0.3.11/lib/python3.9/site-packages/pysyncobj/journal.py", line 258, in createJournal
node2 # [   62.406127] patroni-start[1060]:     return FileJournal(journalFile)
node2 # [   62.408142] patroni-start[1060]:   File "/nix/store/pcpg66fjsn7vah409hynkhirsjphxdya-python3.9-pysyncobj-0.3.11/lib/python3.9/site-packages/pysyncobj/journal.py", line 179, in __init__
node2 # [   62.415208] patroni-start[1060]:     idx, term = struct.unpack('<QQ', nextRecordData[:16])
node2 # [   62.421179] patroni-start[1060]: struct.error: unpack requires a buffer of 16 bytes
```

**Comment by phfroidmont** [comment]
The tests were failing from time to time when using Raft. I couldn't find a solution so I simply used etcd instead. Now it seems reliable and quite a bit faster.
@pennae Thanks a lot for all the feedback so far, it's very helpful.

File: nixos/modules/services/cluster/patroni/default.nix
Comment (pennae) [inline nixos/modules/services/cluster/patroni/default.nix:18]:
```suggestion
    enable = mkEnableOption "Patroni";
```
that prefix is added by `mkEnableOption`

**Review by pennae — COMMENTED** [review]
test succeeds now, looks good over all. just one more little nitpicky thing :)

**Review by pennae — APPROVED** [review]

File: nixos/modules/services/cluster/patroni/default.nix
Comment (pennae) [inline nixos/modules/services/cluster/patroni/default.nix:248]: was about to merge this pr and just gave it a quick once-over. not sure why we didn't say this before, but this part really ought to be done using `boot.kernelModules` and a udev rule that does the `chmod`ing.

Comment (phfroidmont) [reply nixos/modules/services/cluster/patroni/default.nix]: I had a feeling it was wrong when writing it, thanks!

**Pull request merged**
